File: 0020-valid-parentheses/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        par_dict = {
            '(' : ')',
            '[' : ']',
            '{' : '}',
        }
        par_stack = []

        for char in s:
            if char in "([{":
                par_stack.append(char)
            elif len(par_stack) > 0:
                if par_dict[par_stack.pop()] != char:
                    return False
            else:
                return False

        if len(par_stack) > 0:
            return False
        else:
            return True
# @lc code=end

File: 0020-valid-parentheses/20.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
public class Solution {
    public bool IsValid(string s) {
        var parDict = new Dictionary<char, char>(){
            {'(', ')'},
            {'[', ']'},
            {'{', '}'}
        };
        Stack<char> parStack = new Stack<char>();

        foreach (char c in s)
        {
            if ("([{".Contains(c))
            {
                parStack.Push(c);
            }
            else if (parStack.Count > 0)
            {
                if (parDict[parStack.Pop()] != c)
                    return false;
            }
            else return false;
        }

        if (parStack.Count > 0) return false;
        else return true;
    }
}
// @lc code=end

